enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+Actions:
+
+- Create a new repo in OpenShift’s github: https://github.com/openshift/cluster-cloud-controller-manager-operator (Done)
+
+- Implement the operator, using [library-go](https://github.com/openshift/library-go) primitives.
+
+#### Build the operator image by OpenShift automation
+
+To start using the CCM operator in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The image should be automatically tested before it becomes available. Dockerfile should be a part of the operator's repo.
+
+Actions:
+
+- Configure CI operator to build CCM operator image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OpenShift Quay account.
+
+#### Integrate the solution with OpenShift
+
+Actions:
+
+- Make sure the Cinder CSI driver is supported in OpenShift.
+
+- Add CCM operator support in Cluster Version Operator. CCM operator will be deployed on early stages of installation and then it deploys and configures OpenStack CCM itself.
+
+- Change the config observer in [library-go](https://github.com/openshift/library-go/blob/16d6830d0b80dc2a3315207116d009ed2dd4cebf/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L154) to disable setting `--cloud-provider` and `--cloud-config` parameters for OpenStack. Then the library should be bumped in `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` respectively.
+
+- Change `kubelet` configuration for OpenStack in the [MCO](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357) to adopt external cloud providers.
+
+#### Examples for other cloud-controller-manager configurations
+
+- AWS: https://github.com/kubernetes/cloud-provider-aws/
+- Azure: https://github.com/kubernetes-sigs/cloud-provider-azure
+Azure separated the `CCM` and the "cloud-node-manager" (`CNM`) into two binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd.
+- GCP: https://github.com/kubernetes/cloud-provider-gcp
+- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
+vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64
+
+### Risks and Mitigations
+
+#### Specific CCM doesn’t work properly on OpenShift
+
+The `CCMs` have not been tested on OCP yet. Unknown issues may arise.
+
+Severity: medium-high
+Likelihood: medium
+
+#### OpenShift components can't work with the CCM properly
+
+Since OpenShift imposes additional limitations compared to Kubernetes, some collisions are possible.
+So far there are no CCMs available in OpenShift, and there can be some problems with operability.
+
+Severity: medium-high
+Likelihood: low
+
+#### CSI driver is not available in OpenShift
```

Comment:
I have no idea how CSI drivers fall in the big picture of out-of-tree providers, it wasn't covered earlier.

Reply:
Does the updated content below help or should we be mentioning this earlier in the proposal?